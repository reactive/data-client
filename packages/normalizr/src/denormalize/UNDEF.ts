export const UNDEF = {} as any;
