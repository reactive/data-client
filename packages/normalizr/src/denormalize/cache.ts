import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
