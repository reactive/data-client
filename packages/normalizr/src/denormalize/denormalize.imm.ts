import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';
import { MemoPolicy } from '../memo/Policy.imm.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    MemoPolicy.getEntities(entities),
    new LocalCache(),
    args,
  )(schema, input).data;
}
