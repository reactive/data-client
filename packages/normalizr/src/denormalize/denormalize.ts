import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
