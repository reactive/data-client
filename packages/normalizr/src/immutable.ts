export { DelegateImmutable } from './memo/Delegate.immutable.js';
