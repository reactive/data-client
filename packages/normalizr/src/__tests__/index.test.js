// eslint-env jest
import { Entity, schema } from '@data-client/endpoint';
import { fromJS } from 'immutable';

import { normalize } from '../';
import { denormalize as denormalizeSimple } from '../denormalize/denormalize';
import { INVALID } from '../denormalize/symbol';
import MemoCached from '../memo/MemoCache';

class IDEntity extends Entity {
  id = '';
  pk(parent, key) {
    return this.id || key;
  }
}

class Tacos extends IDEntity {
  type = '';
}

let dateSpy;
beforeAll(() => {
  dateSpy = jest
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe('normalize', () => {
  test.each([42, null, undefined, '42', () => {}])(
    `cannot normalize input that == %s`,
    input => {
      class Test extends IDEntity {}
      expect(() => normalize(Test, input)).toThrow();
    },
  );
  test.each([42, null, undefined, '42', () => {}])(
    `cannot normalize input that == %s`,
    input => {
      class Test extends IDEntity {}
      expect(() => normalize({ data: Test }, input)).toThrow();
    },
  );

  test('can normalize strings for entity (already processed)', () => {
    class Test extends IDEntity {}
    expect(normalize(new schema.Array(Test), ['42']).result).toEqual(['42']);
  });

  test('can normalize null schema with string response', () => {
    expect(normalize(null, '17,234').result).toEqual('17,234');
  });

  test('passthrough with undefined schema', () => {
    const input = {};
    expect(normalize(undefined, input).result).toBe(input);
  });

  test('passthrough with id in place of entity', () => {
    const input = { taco: '5' };
    expect(normalize({ taco: Tacos }, input).result).toStrictEqual(input);
  });

  test('cannot normalize with null input', () => {
    expect(() => normalize(Tacos, null)).toThrow(/null/);
  });

  test('passthrough primitive schema', () => {
    expect(normalize({ happy: 5 }, { happy: { bob: 5 } }).result).toStrictEqual(
      {
        happy: { bob: 5 },
      },
    );
  });

  test('can normalize string', () => {
    const mySchema = '';
    expect(normalize(mySchema, 'bob')).toMatchInlineSnapshot(`
      {
        "entities": {},
        "entityMeta": {},
        "indexes": {},
        "result": "bob",
      }
    `);
  });

  test('normalizes entities', () => {
    expect(
      normalize(
        [Tacos],
        [
          { id: '1', type: 'foo' },
          { id: '2', type: 'bar' },
        ],
      ),
    ).toMatchSnapshot();
  });

  test('normalizes schema with extra members', () => {
    expect(
      normalize(
        {
          data: [Tacos],
          extra: '',
          page: {
            first: null,
            second: undefined,
            third: 0,
            complex: { complex: true, next: false },
          },
        },
        {
          data: [
            { id: '1', type: 'foo' },
            { id: '2', type: 'bar' },
          ],
          extra: 'five',
          page: {
            first: null,
            second: { thing: 'two' },
            third: 1,
            complex: { complex: false, next: true },
          },
        },
      ),
    ).toMatchSnapshot();
  });

  test('normalizes schema with extra members but not set', () => {
    expect(
      normalize(
        {
          data: [Tacos],
          extra: '',
          page: {
            first: null,
            second: undefined,
            third: 0,
            complex: { complex: true, next: false },
          },
        },
        {
          data: [
            { id: '1', type: 'foo' },
            { id: '2', type: 'bar' },
          ],
        },
      ),
    ).toMatchSnapshot();
  });

  test('normalizes schema with indexes', () => {
    class MyTaco extends Tacos {
      static indexes = ['type'];
    }

    expect(
      normalize(
        { data: [MyTaco], alt: MyTaco },
        {
          data: [
            { id: '1', type: 'foo' },
            { id: '2', type: 'bar' },
          ],
          alt: { id: '2', type: 'bar2' },
        },
      ),
    ).toMatchSnapshot();
  });

  test('normalizes warns on schemas with unfound indexes', () => {
    const oldError = console.warn;
    const spy = (console.warn = jest.fn());

    class MyTaco extends Tacos {
      static indexes = ['notfound'];
    }

    expect(
      normalize(
        { data: [MyTaco], alt: MyTaco },
        {
          data: [
            { id: '1', type: 'foo' },
            { id: '2', type: 'bar' },
          ],
          alt: { id: '2', type: 'bar2' },
        },
      ),
    ).toMatchSnapshot();

    expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "Index not found in entity. Indexes must be top-level members of your entity.
      Index: notfound
      Entity: {
        "id": "1",
        "type": "foo"
      }",
      ]
    `);
    console.warn = oldError;
  });

  test('normalizes entities with circular references', () => {
    class User extends IDEntity {}
    User.schema = { friends: [User] };

    const input = { id: '123', friends: [] };
    input.friends.push(input);

    expect(normalize(User, input)).toMatchSnapshot();
  });

  test('normalizes entities with circular references that fails validation', () => {
    class User extends IDEntity {
      static validate(processedEntity) {
        return 'this always fails';
      }
    }
    User.schema = { friends: [User] };

    const input = { id: '123', friends: [] };
    input.friends.push(input);

    expect(() => normalize(User, input)).toThrowErrorMatchingSnapshot();
  });

  test('normalizes nested entities', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      static schema = {
        author: User,
        comments: [Comment],
      };
    }

    const input = {
      id: '123',
      title: 'A Great Article',
      author: {
        id: '8472',
        name: 'Paul',
      },
      body: 'This article is great.',
      comments: [
        {
          id: 'comment-123-4738',
          comment: 'I like it!',
          user: {
            id: '10293',
            name: 'Jane',
          },
        },
      ],
    };
    expect(normalize(Article, input)).toMatchSnapshot();
  });

  test('does not modify the original input', () => {
    class User extends IDEntity {}
    class Article extends IDEntity {
      title = '';
      static schema = {
        author: User,
      };
    }
    const input = Object.freeze({
      id: '123',
      title: 'A Great Article',
      author: Object.freeze({
        id: '8472',
        name: 'Paul',
      }),
    });
    expect(() => normalize(Article, input)).not.toThrow();
  });

  test('handles number ids when nesting', () => {
    class User extends IDEntity {}
    class Article extends IDEntity {
      title = '';
      static schema = {
        author: User,
      };
    }
    const input = Object.freeze({
      id: 123,
      title: 'A Great Article',
      author: {
        id: 8472,
        name: 'Paul',
      },
    });
    expect(normalize(Article, input).entities).toMatchSnapshot();
  });

  test('ignores null values', () => {
    class MyEntity extends IDEntity {}
    expect(normalize([MyEntity], [null])).toMatchSnapshot();
    expect(normalize([MyEntity], [undefined])).toMatchSnapshot();
    expect(normalize([MyEntity], [false])).toMatchSnapshot();
  });

  test('can use fully custom entity classes', () => {
    class Children extends IDEntity {
      name = '';
    }
    class MyEntity extends Entity {
      uuid = '';
      name = '';

      static schema = {
        children: [Children],
      };

      pk(parent, key) {
        return this.uuid;
      }

      static normalize(
        input,
        parent,
        key,
        args,
        visit,
        addEntity,
        getEntity,
        checkLoop,
      ) {
        const entity = { ...input };
        Object.keys(this.schema).forEach(key => {
          const schema = this.schema[key];
          entity[key] = visit(schema, input[key], input, key, args);
        });
        addEntity(this, entity, this.pk(entity));
        return {
          uuid: this.pk(entity),
          schema: this.key,
        };
      }
    }

    class Food extends MyEntity {}
    expect(
      normalize(Food, {
        uuid: '1234',
        name: 'tacos',
        children: [{ id: 4, name: 'lettuce' }],
      }),
    ).toMatchSnapshot();
  });

  test('uses the non-normalized input when getting the ID for an entity', () => {
    const calls = [];
    class User extends IDEntity {}
    class Recommendations extends Entity {
      pk(parent, key) {
        calls.push([this, parent, key]);
        return parent.user.id;
      }

      static schema = { user: User };
    }

    expect(
      normalize(Recommendations, { user: { id: '456' } }, [{ id: '456' }]),
    ).toMatchSnapshot();
    expect(calls).toMatchSnapshot();
  });

  test('passes over pre-normalized values', () => {
    class User extends IDEntity {}
    class Article extends IDEntity {
      title = '';
      static schema = { author: User };
    }

    expect(
      normalize(Article, {
        id: '123',
        title: 'normalizr is great!',
        author: '1',
      }),
    ).toMatchSnapshot();

    expect(normalize({ user: User }, { user: '1' })).toMatchSnapshot();
  });

  test('can normalize object without proper object prototype inheritance', () => {
    const test = { id: '1', elements: [] };
    test.elements.push(
      Object.assign(Object.create(null), {
        id: '18',
        name: 'test',
      }),
    );

    class Elements extends IDEntity {}
    class Test extends IDEntity {
      static schema = {
        elements: [Elements],
      };
    }

    expect(() => normalize(Test, test)).not.toThrow();
  });
});

const memo = new MemoCached();
function denormalizeCachedValue(...args) {
  return memo.denormalize(...args).data;
}
describe.each([
  ['fast', denormalizeSimple],
  ['cached', denormalizeCachedValue],
])(`denormalize [%s]`, (_, denormalize) => {
  test('passthrough with undefined schema', () => {
    const input = {};
    expect(denormalize(undefined, input)).toEqual(input);
  });

  test('returns the input if undefined', () => {
    expect(denormalize({}, undefined, {})).toEqual(undefined);
  });

  test('returns the input if string', () => {
    expect(denormalize('', 'bob', {})).toEqual('bob');
  });

  test('denormalizes entities', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
        2: { id: '2', type: 'bar' },
      },
    };
    expect(denormalize([Tacos], ['1', '2'], entities)).toMatchSnapshot();
  });

  test('denormalizes without entities fills undefined', () => {
    expect(denormalize({ data: Tacos }, { data: '1' }, {})).toMatchSnapshot();
    expect(
      denormalize({ data: Tacos }, fromJS({ data: '1' }), {}),
    ).toMatchSnapshot();
    expect(denormalize(Tacos, '1', {})).toEqual(undefined);
  });

  test('denormalizes ignoring unfound entities in arrays', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
      },
    };
    expect(denormalize([Tacos], ['1', '2'], entities)).toMatchSnapshot();
    expect(
      denormalize({ results: [Tacos] }, { results: ['1', '2'] }, entities),
    ).toMatchSnapshot();
  });

  test('denormalizes suspends when symbol contains INVALID string', () => {
    const entities = {
      Tacos: {
        1: Symbol('ENTITY WAS INVALID'),
      },
    };
    expect(denormalize(Tacos, '1', entities)).toEqual(expect.any(Symbol));
  });

  test('denormalizes ignoring deleted entities in arrays', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
        2: INVALID,
      },
    };
    expect(denormalize([Tacos], ['1', '2'], entities)).toMatchSnapshot();
    expect(
      denormalize({ results: [Tacos] }, { results: ['1', '2'] }, entities),
    ).toMatchSnapshot();
  });

  test('denormalizes arrays with objects inside', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
      },
    };
    /*expect(
      denormalize([{ data: Tacos }],[{ data: 1 }, { data: 2 }],  {}),
    ).toEqual([]);*/
    expect(
      denormalize([{ data: Tacos }], [{ data: 1 }, { data: 2 }], entities),
    ).toMatchSnapshot();
  });

  test('denormalizes schema with extra members', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
        2: { id: '2', type: 'bar' },
      },
    };
    expect(
      denormalize(
        {
          data: [Tacos],
          extra: '',
          page: {
            first: null,
            second: undefined,
            third: 0,
            complex: { complex: true, next: false },
          },
        },
        {
          data: ['1', '2'],
          extra: '5',
          page: {
            first: null,
            second: { thing: 'two' },
            third: 1,
            complex: { complex: false, next: true },
          },
        },

        entities,
      ),
    ).toMatchSnapshot();
  });

  test('denormalizes schema with extra members but not set', () => {
    const entities = {
      Tacos: {
        1: { id: '1', type: 'foo' },
        2: { id: '2', type: 'bar' },
      },
    };
    expect(
      denormalize(
        {
          data: [Tacos],
          extra: '',
          page: {
            first: null,
            second: undefined,
            third: 0,
            complex: { complex: true, next: false },
          },
        },
        {
          data: ['1', '2'],
        },

        entities,
      ),
    ).toMatchSnapshot();
  });

  test('denormalizes nested entities', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      static schema = {
        author: User,
        comments: [Comment],
      };
    }

    const entities = {
      Article: {
        123: {
          author: '8472',
          body: 'This article is great.',
          comments: ['comment-123-4738'],
          id: '123',
          title: 'A Great Article',
        },
      },
      Comment: {
        'comment-123-4738': {
          comment: 'I like it!',
          id: 'comment-123-4738',
          user: '10293',
        },
      },
      User: {
        10293: {
          id: '10293',
          name: 'Jane',
        },
        8472: {
          id: '8472',
          name: 'Paul',
        },
      },
    };
    expect(denormalize(Article, '123', entities)).toMatchSnapshot();
  });

  test('gracefully handles when nested entities are primitives', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      static schema = {
        author: User,
        comments: [Comment],
      };
    }

    const entities = {
      Article: {
        123: {
          author: '8472',
          body: 'This article is great.',
          comments: ['comment-123-4738'],
          id: '123',
          title: 'A Great Article',
        },
      },
      Comment: {
        'comment-123-4738': 4,
      },
      User: {
        10293: {
          id: '10293',
          name: 'Jane',
        },
        8472: {
          id: '8472',
          name: 'Paul',
        },
      },
    };
    expect(() => denormalize(Article, '123', entities)).toMatchSnapshot();
  });

  test('set to undefined if schema key is not in entities', () => {
    class User extends IDEntity {}
    class Comment extends IDEntity {
      comment = '';
      static schema = {
        user: User,
      };
    }
    class Article extends IDEntity {
      title = '';
      body = '';
      static schema = {
        author: User,
        comments: [Comment],
      };

      static validate() {}
    }

    const entities = {
      Article: {
        123: {
          id: '123',
          author: '8472',
          comments: ['1'],
        },
      },
      Comment: {
        1: {
          id: '1',
          user: '123',
        },
      },
    };
    expect(denormalize(Article, '123', entities)).toMatchSnapshot();
  });

  test('does not modify the original entities', () => {
    class User extends IDEntity {}
    class Article extends IDEntity {
      static schema = {
        author: User,
      };
    }
    const entities = Object.freeze({
      Article: Object.freeze({
        123: Object.freeze({
          id: '123',
          title: 'A Great Article',
          author: '8472',
        }),
      }),
      User: Object.freeze({
        8472: Object.freeze({
          id: '8472',
          name: 'Paul',
        }),
      }),
    });
    expect(() => denormalize(Article, '123', entities)).not.toThrow();
  });

  test('denormalizes with function as pk()', () => {
    class Guest extends Entity {
      pk(parent, key) {
        return `${key}-${parent.id}-${this.guestId}`;
      }
    }
    class Patron extends IDEntity {
      static schema = {
        guest: Guest,
      };
    }

    const normalizedData = {
      entities: {
        Patron: {
          1: { id: '1', guest: null, name: 'Esther' },
          2: { id: '2', guest: 'guest-2-1', name: 'Tom' },
        },
        Guest: { 'guest-2-1': { guestId: 1 } },
      },
      result: ['1', '2'],
    };

    expect(
      denormalize([Patron], normalizedData.result, normalizedData.entities),
    ).toMatchSnapshot();
  });

  test('denormalizes where id is only in key', () => {
    expect(
      denormalize(
        new schema.Values(Tacos),
        {
          1: { type: 'foo' },
          2: { type: 'bar' },
        },
        {},
      ),
    ).toMatchSnapshot();
  });
});
