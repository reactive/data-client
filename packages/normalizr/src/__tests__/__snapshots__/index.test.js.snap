// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`denormalize [cached] denormalizes arrays with objects inside 1`] = `
[
  {
    "data": Tacos {
      "id": "1",
      "type": "foo",
    },
  },
  {
    "data": undefined,
  },
]
`;

exports[`denormalize [cached] denormalizes entities 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
]
`;

exports[`denormalize [cached] denormalizes ignoring deleted entities in arrays 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`denormalize [cached] denormalizes ignoring deleted entities in arrays 2`] = `
{
  "results": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
  ],
}
`;

exports[`denormalize [cached] denormalizes ignoring unfound entities in arrays 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`denormalize [cached] denormalizes ignoring unfound entities in arrays 2`] = `
{
  "results": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
  ],
}
`;

exports[`denormalize [cached] denormalizes nested entities 1`] = `
Article {
  "author": User {
    "id": "8472",
    "name": "Paul",
  },
  "body": "This article is great.",
  "comments": [
    Comment {
      "comment": "I like it!",
      "id": "comment-123-4738",
      "user": User {
        "id": "10293",
        "name": "Jane",
      },
    },
  ],
  "id": "123",
  "title": "A Great Article",
}
`;

exports[`denormalize [cached] denormalizes schema with extra members 1`] = `
{
  "data": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
    Tacos {
      "id": "2",
      "type": "bar",
    },
  ],
  "extra": "5",
  "page": {
    "complex": {
      "complex": false,
      "next": true,
    },
    "first": null,
    "second": {
      "thing": "two",
    },
    "third": 1,
  },
}
`;

exports[`denormalize [cached] denormalizes schema with extra members but not set 1`] = `
{
  "data": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
    Tacos {
      "id": "2",
      "type": "bar",
    },
  ],
}
`;

exports[`denormalize [cached] denormalizes where id is only in key 1`] = `
{
  "1": Tacos {
    "id": "",
    "type": "foo",
  },
  "2": Tacos {
    "id": "",
    "type": "bar",
  },
}
`;

exports[`denormalize [cached] denormalizes with function as pk() 1`] = `
[
  Patron {
    "guest": null,
    "id": "1",
    "name": "Esther",
  },
  Patron {
    "guest": Guest {
      "guestId": 1,
    },
    "id": "2",
    "name": "Tom",
  },
]
`;

exports[`denormalize [cached] denormalizes without entities fills undefined 1`] = `
{
  "data": undefined,
}
`;

exports[`denormalize [cached] denormalizes without entities fills undefined 2`] = `
Immutable.Map {
  "data": undefined,
}
`;

exports[`denormalize [cached] gracefully handles when nested entities are primitives 1`] = `[Function]`;

exports[`denormalize [cached] set to undefined if schema key is not in entities 1`] = `
Article {
  "author": undefined,
  "body": "",
  "comments": [
    Comment {
      "comment": "",
      "id": "1",
      "user": undefined,
    },
  ],
  "id": "123",
  "title": "",
}
`;

exports[`denormalize [fast] denormalizes arrays with objects inside 1`] = `
[
  {
    "data": Tacos {
      "id": "1",
      "type": "foo",
    },
  },
  {
    "data": undefined,
  },
]
`;

exports[`denormalize [fast] denormalizes entities 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
  Tacos {
    "id": "2",
    "type": "bar",
  },
]
`;

exports[`denormalize [fast] denormalizes ignoring deleted entities in arrays 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`denormalize [fast] denormalizes ignoring deleted entities in arrays 2`] = `
{
  "results": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
  ],
}
`;

exports[`denormalize [fast] denormalizes ignoring unfound entities in arrays 1`] = `
[
  Tacos {
    "id": "1",
    "type": "foo",
  },
]
`;

exports[`denormalize [fast] denormalizes ignoring unfound entities in arrays 2`] = `
{
  "results": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
  ],
}
`;

exports[`denormalize [fast] denormalizes nested entities 1`] = `
Article {
  "author": User {
    "id": "8472",
    "name": "Paul",
  },
  "body": "This article is great.",
  "comments": [
    Comment {
      "comment": "I like it!",
      "id": "comment-123-4738",
      "user": User {
        "id": "10293",
        "name": "Jane",
      },
    },
  ],
  "id": "123",
  "title": "A Great Article",
}
`;

exports[`denormalize [fast] denormalizes schema with extra members 1`] = `
{
  "data": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
    Tacos {
      "id": "2",
      "type": "bar",
    },
  ],
  "extra": "5",
  "page": {
    "complex": {
      "complex": false,
      "next": true,
    },
    "first": null,
    "second": {
      "thing": "two",
    },
    "third": 1,
  },
}
`;

exports[`denormalize [fast] denormalizes schema with extra members but not set 1`] = `
{
  "data": [
    Tacos {
      "id": "1",
      "type": "foo",
    },
    Tacos {
      "id": "2",
      "type": "bar",
    },
  ],
}
`;

exports[`denormalize [fast] denormalizes where id is only in key 1`] = `
{
  "1": Tacos {
    "id": "",
    "type": "foo",
  },
  "2": Tacos {
    "id": "",
    "type": "bar",
  },
}
`;

exports[`denormalize [fast] denormalizes with function as pk() 1`] = `
[
  Patron {
    "guest": null,
    "id": "1",
    "name": "Esther",
  },
  Patron {
    "guest": Guest {
      "guestId": 1,
    },
    "id": "2",
    "name": "Tom",
  },
]
`;

exports[`denormalize [fast] denormalizes without entities fills undefined 1`] = `
{
  "data": undefined,
}
`;

exports[`denormalize [fast] denormalizes without entities fills undefined 2`] = `
Immutable.Map {
  "data": undefined,
}
`;

exports[`denormalize [fast] gracefully handles when nested entities are primitives 1`] = `[Function]`;

exports[`denormalize [fast] set to undefined if schema key is not in entities 1`] = `
Article {
  "author": undefined,
  "body": "",
  "comments": [
    Comment {
      "comment": "",
      "id": "1",
      "user": undefined,
    },
  ],
  "id": "123",
  "title": "",
}
`;

exports[`normalize can use fully custom entity classes 1`] = `
{
  "entities": {
    "Children": {
      "4": {
        "id": 4,
        "name": "lettuce",
      },
    },
    "Food": {
      "1234": {
        "children": [
          "4",
        ],
        "name": "tacos",
        "uuid": "1234",
      },
    },
  },
  "entitiesMeta": {
    "Children": {
      "4": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Food": {
      "1234": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "schema": "Food",
    "uuid": "1234",
  },
}
`;

exports[`normalize handles number ids when nesting 1`] = `
{
  "Article": {
    "123": {
      "author": "8472",
      "id": 123,
      "title": "A Great Article",
    },
  },
  "User": {
    "8472": {
      "id": 8472,
      "name": "Paul",
    },
  },
}
`;

exports[`normalize ignores null values 1`] = `
{
  "entities": {},
  "entitiesMeta": {},
  "indexes": {},
  "result": [
    null,
  ],
}
`;

exports[`normalize ignores null values 2`] = `
{
  "entities": {},
  "entitiesMeta": {},
  "indexes": {},
  "result": [
    undefined,
  ],
}
`;

exports[`normalize ignores null values 3`] = `
{
  "entities": {},
  "entitiesMeta": {},
  "indexes": {},
  "result": [
    false,
  ],
}
`;

exports[`normalize normalizes entities 1`] = `
{
  "entities": {
    "Tacos": {
      "1": {
        "id": "1",
        "type": "foo",
      },
      "2": {
        "id": "2",
        "type": "bar",
      },
    },
  },
  "entitiesMeta": {
    "Tacos": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "1",
    "2",
  ],
}
`;

exports[`normalize normalizes entities with circular references 1`] = `
{
  "entities": {
    "User": {
      "123": {
        "friends": [
          "123",
        ],
        "id": "123",
      },
    },
  },
  "entitiesMeta": {
    "User": {
      "123": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "123",
}
`;

exports[`normalize normalizes entities with circular references that fails validation 1`] = `"this always fails"`;

exports[`normalize normalizes nested entities 1`] = `
{
  "entities": {
    "Article": {
      "123": {
        "author": "8472",
        "body": "This article is great.",
        "comments": [
          "comment-123-4738",
        ],
        "id": "123",
        "title": "A Great Article",
      },
    },
    "Comment": {
      "comment-123-4738": {
        "comment": "I like it!",
        "id": "comment-123-4738",
        "user": "10293",
      },
    },
    "User": {
      "10293": {
        "id": "10293",
        "name": "Jane",
      },
      "8472": {
        "id": "8472",
        "name": "Paul",
      },
    },
  },
  "entitiesMeta": {
    "Article": {
      "123": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "Comment": {
      "comment-123-4738": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "10293": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "8472": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "123",
}
`;

exports[`normalize normalizes schema with extra members 1`] = `
{
  "entities": {
    "Tacos": {
      "1": {
        "id": "1",
        "type": "foo",
      },
      "2": {
        "id": "2",
        "type": "bar",
      },
    },
  },
  "entitiesMeta": {
    "Tacos": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "data": [
      "1",
      "2",
    ],
    "extra": "five",
    "page": {
      "complex": {
        "complex": false,
        "next": true,
      },
      "first": null,
      "second": {
        "thing": "two",
      },
      "third": 1,
    },
  },
}
`;

exports[`normalize normalizes schema with extra members but not set 1`] = `
{
  "entities": {
    "Tacos": {
      "1": {
        "id": "1",
        "type": "foo",
      },
      "2": {
        "id": "2",
        "type": "bar",
      },
    },
  },
  "entitiesMeta": {
    "Tacos": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": {
    "data": [
      "1",
      "2",
    ],
  },
}
`;

exports[`normalize normalizes schema with indexes 1`] = `
{
  "entities": {
    "MyTaco": {
      "1": {
        "id": "1",
        "type": "foo",
      },
      "2": {
        "id": "2",
        "type": "bar2",
      },
    },
  },
  "entitiesMeta": {
    "MyTaco": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {
    "MyTaco": {
      "type": {
        "bar": Symbol(INVALID),
        "bar2": "2",
        "foo": "1",
      },
    },
  },
  "result": {
    "alt": "2",
    "data": [
      "1",
      "2",
    ],
  },
}
`;

exports[`normalize normalizes warns on schemas with unfound indexes 1`] = `
{
  "entities": {
    "MyTaco": {
      "1": {
        "id": "1",
        "type": "foo",
      },
      "2": {
        "id": "2",
        "type": "bar2",
      },
    },
  },
  "entitiesMeta": {
    "MyTaco": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "2": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {
    "MyTaco": {
      "notfound": {},
    },
  },
  "result": {
    "alt": "2",
    "data": [
      "1",
      "2",
    ],
  },
}
`;

exports[`normalize passes over pre-normalized values 1`] = `
{
  "entities": {
    "Article": {
      "123": {
        "author": "1",
        "id": "123",
        "title": "normalizr is great!",
      },
    },
  },
  "entitiesMeta": {
    "Article": {
      "123": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "123",
}
`;

exports[`normalize passes over pre-normalized values 2`] = `
{
  "entities": {},
  "entitiesMeta": {},
  "indexes": {},
  "result": {
    "user": "1",
  },
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 1`] = `
{
  "entities": {
    "Recommendations": {
      "456": {
        "user": "456",
      },
    },
    "User": {
      "456": {
        "id": "456",
      },
    },
  },
  "entitiesMeta": {
    "Recommendations": {
      "456": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "456": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "456",
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 2`] = `
[
  [
    {
      "user": "456",
    },
    {
      "user": {
        "id": "456",
      },
    },
    undefined,
  ],
]
`;
